File: src/main/java/com/nuutrai/reactor/commands/RegisterCommands.java
package com.nuutrai.reactor.commands;

import com.mojang.brigadier.Command;
import com.mojang.brigadier.arguments.ArgumentType;
import com.mojang.brigadier.arguments.IntegerArgumentType;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.builder.ArgumentBuilder;
import com.nuutrai.reactor.Reactor;
import com.nuutrai.reactor.data.DataManager;
import com.nuutrai.reactor.player.PlayerData;
import com.nuutrai.reactor.store.Store;
import io.papermc.paper.command.brigadier.Commands;
import io.papermc.paper.plugin.lifecycle.event.LifecycleEventManager;
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents;
import org.bukkit.Warning;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;
import java.util.List;

import static com.nuutrai.reactor.Reactor.logger;
import static io.papermc.paper.command.brigadier.Commands.argument;

@SuppressWarnings("UnstableApiUsage")
public class RegisterCommands {

    public static void loadInventoryTest() {
        Reactor.manager.registerEventHandler(LifecycleEvents.COMMANDS, event -> {
            final Commands commands = event.registrar();
            commands.register(
                    Commands.literal("testInv")
                            .executes(ctx -> {
                                if (ctx.getSource().getSender() instanceof Player player) {
                                    Store.setup(player);
                                    return Command.SINGLE_SUCCESS;
                                }
                                return 0;
                            })
                            .build()
            );
        });
    }

    @SuppressWarnings("NullableProblems")
    @Deprecated
        public static void loadDataEditor() {
            LifecycleEventManager<Plugin> manager = Reactor.instance.getLifecycleManager();
            manager.registerEventHandler(LifecycleEvents.COMMANDS, event -> {
                final Commands commands = event.registrar();
                commands.register(
                        Commands.literal("setData")
                                .then(argument("name", StringArgumentType.word())
                                        .then(argument("value", StringArgumentType.greedyString())
                                            .executes(ctx -> {
                                                CommandSender sender = ctx.getSource().getSender();
                                                if (!(sender instanceof Player player))
                                                    return 0;

                                                PlayerData playerData = DataManager.get(player);
                                                String name = StringArgumentType.getString(ctx, "name");
                                                String value = StringArgumentType.getString(ctx, "value");

                                                logger.info("Hello??");
                                                logger.info(playerData.toString());
                                                logger.info(name);
                                                logger.info(value);

                                                try {
                                                    PlayerData.class.getMethod(name, String.class).invoke(playerData, value);
                                                    return Command.SINGLE_SUCCESS;
                                                } catch (IllegalAccessException | NoSuchMethodException |
                                                         InvocationTargetException e) {
                                                    throw new RuntimeException(e);
                                                }

                                            }))
                                        .executes(ctx -> {
                                            ctx.getSource().getSender().sendMessage("Uh, what was that?");
                                            return 0;
                                        }))
                                .executes(ctx -> {
                                    ctx.getSource().getSender().sendMessage("Uh, what was that?");
                                    return 0;
                                })
                                .build(),
                        "some bukkit help description string",
                        List.of("an-alias")
                );
            });
        }

}

File: src/main/kotlin/com/nuutrai/reactor/commands/RegisterCommands.kt
@file:Suppress("UnstableApiUsage")

package com.nuutrai.reactor.commands

import com.mojang.brigadier.Command
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.context.CommandContext
import com.nuutrai.reactor.Reactor
import com.nuutrai.reactor.data.DataManager.get
import com.nuutrai.reactor.player.PlayerData
import com.nuutrai.reactor.store.Store.setup
import io.papermc.paper.command.brigadier.CommandSourceStack
import io.papermc.paper.command.brigadier.Commands
import io.papermc.paper.plugin.lifecycle.event.handler.LifecycleEventHandler
import io.papermc.paper.plugin.lifecycle.event.registrar.ReloadableRegistrarEvent
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents
import org.bukkit.entity.Player
import java.lang.Deprecated
import java.lang.reflect.InvocationTargetException
import kotlin.RuntimeException
import kotlin.String
import kotlin.Suppress
import kotlin.Throws
import kotlin.toString

object RegisterCommands {
	@Throws(InvocationTargetException::class, IllegalAccessException::class)
	fun load() {
		val methods = RegisterCommands::class.java.getDeclaredMethods()
		for (m in methods) {
			if (!m.isAnnotationPresent(Deprecated::class.java) && (m.name != "load") && !m.name
					.contains("$")
			) {
				m.invoke(null)
			}
		}
	}

	fun loadInventoryTest() {
		Reactor.manager!!.registerEventHandler<ReloadableRegistrarEvent<Commands>>(
			LifecycleEvents.COMMANDS,
			LifecycleEventHandler { event: ReloadableRegistrarEvent<Commands>? ->
				val commands = event!!.registrar()
				commands.register(
					Commands.literal("testInv")
						.executes(Command { ctx: CommandContext<CommandSourceStack?>? ->
							setup((ctx!!.getSource()!!.sender as? Player) ?: return@Command 0)
							return@Command Command.SINGLE_SUCCESS
						})
						.build()
				)
			})
	}

	@kotlin.Deprecated("")
	fun loadDataEditor() {
		val manager = Reactor.instance!!.lifecycleManager
		manager.registerEventHandler<ReloadableRegistrarEvent<Commands>>(
			LifecycleEvents.COMMANDS,
			LifecycleEventHandler { event: ReloadableRegistrarEvent<Commands>? ->
				val commands = event!!.registrar()
				commands.register(
					Commands.literal("setData")
						.then(
							Commands.argument<String>("name", StringArgumentType.word()!!)
								.then(
									Commands.argument<String>("value", StringArgumentType.greedyString()!!)
										.executes(Command { ctx: CommandContext<CommandSourceStack?>? ->
											val sender = ctx!!.getSource()!!.sender
											if (sender !is Player) return@Command 0

											val playerData = get(sender)
											val name = StringArgumentType.getString(ctx, "name")
											val value = StringArgumentType.getString(ctx, "value")

											Reactor.Companion.logger!!.info("Hello??")
											Reactor.Companion.logger!!.info(playerData.toString())
											Reactor.Companion.logger!!.info(name)
											Reactor.Companion.logger!!.info(value)
											try {
												PlayerData::class.java.getMethod(name, String::class.java)
													.invoke(playerData, value)
												return@Command Command.SINGLE_SUCCESS
											} catch (e: IllegalAccessException) {
												throw RuntimeException(e)
											} catch (e: NoSuchMethodException) {
												throw RuntimeException(e)
											} catch (e: InvocationTargetException) {
												throw RuntimeException(e)
											}
										})
								)
								.executes { ctx: CommandContext<CommandSourceStack?>? ->
									ctx!!.getSource()!!.sender.sendMessage("Uh, what was that?")
									0
								}
						)
						.executes { ctx: CommandContext<CommandSourceStack?>? ->
							ctx!!.getSource()!!.sender.sendMessage("Uh, what was that?")
							0
						}
						.build(),
					"some bukkit help description string",
					mutableListOf("an-alias")
				)
			})
	}
}
